### PR TITLE
[Core] Trigger Failover When Error Raised in Creation Task

### PR DESCRIPTION
**Pull request by lixin-wei**
## Why are these changes needed?

Now if an error is raised in actor's constructor, actor will remember this error and reraise this error in incoming tasks.

However, this makes the failover behavior of actor to be strange, some user is wondering: "why my actor was created failed but not been restarted? I've set max_restarts. Instead, it hangs there and keeps raising errors when I call it."

These users often write some codes that contain uncertain behaviors in the constructor like establishing a network connection, when a network error occurs, they want the actor to be restarted and try again until max_restarts exceed.

So the main problem is how we define 'when to rigger actor failover'(when max_restarts not exceed).

Now it is: 
1. process exits.

I want to change it to:
1. error raised in the constructor.
2. process exits.

What do you think?

This PR makes actors to be restarted when error raised in the constructor.

**Side effect that might happen after this PR:**
1.  User can't get init error by calling this actor anymore. They can only get this error by events.
2. If user set a huge max_restars and wirte bad codes in the constructor, the scale of chaos will be larger after this PR, because we'll keep restarting this actor. But I think it's ok because it's not our responsibility to deal with this.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/worker.py [hunk 1]
```diff
@@ -162,17 +162,6 @@ def current_session_and_job(self):
         assert isinstance(self.current_job_id, ray.JobID)
         return self._session_index, self.current_job_id
 
-    def mark_actor_init_failed(self, error):
```
Comment (kfstorm) [inline python/ray/worker.py:165]: Please also remove `self.actor_init_error = None`

File: java/test/src/main/java/io/ray/test/FailureTest.java [hunk 2]
```diff
@@ -75,6 +82,29 @@ public int badMethod2() {
...
+      File flagFile = new File(flagFileName);
+      if (!flagFile.exists()) {
+        LOGGER.info("flag file not exists, throw exception!, fileName={}", flagFile.toString());
+        try {
+          flagFile.createNewFile();
+        } catch (IOException e) {
+          throw new RuntimeException(e);
+        }
+        throw new RuntimeException(EXCEPTION_MESSAGE);
+      }
```
Comment (kfstorm) [inline java/test/src/main/java/io/ray/test/FailureTest.java:100]: Use `Ray.getRuntimeContext().wasCurrentActorRestarted()` to simplify this.

File: python/ray/tests/test_failure.py [hunk 3]
```diff
@@ -291,6 +292,44 @@ def fail_method(self):
...
+
+    @ray.remote(max_restarts=2)
+    class FailedActor:
+        def __init__(self):
+            # This actor will raise error at first time.
+            # Then after restarting, it will be created successfully
+            if not os.path.isfile(error_flag_file_name):
+                open(error_flag_file_name, "w").close()
+                print("flag not exits, raise error")
+                raise Exception(error_message1)
```
Comment (kfstorm) [inline python/ray/tests/test_failure.py:309]: Use `ray.get_runtime_context().was_current_actor_reconstructed` to simplify this. And could you please also help rename `was_current_actor_reconstructed` to `was_current_actor_restarted`?

File: java/test/src/main/java/io/ray/test/FailureTest.java [hunk 4]
```diff
@@ -98,6 +128,27 @@ public void testActorCreationFailure() {
...
+    String flagFileName = "/tmp/TEST_ERROR_FLAG_" + new Random().nextInt(9999);
+    ActorHandle<ActorFailedAtFirstTime> actor =
+        Ray.actor(ActorFailedAtFirstTime::new, flagFileName).setMaxRestarts(2).remote();
+    // At first, actor will throw exceptions and restarted.
+    // Finally after restarted, actor will be created successfully.
+    boolean ok =
+        TestUtils.waitForCondition(
+            () -> {
+              try {
+                Assert.assertEquals(actor.task(ActorFailedAtFirstTime::ok).remote().get(), "OK");
```
Comment (kfstorm) [inline java/test/src/main/java/io/ray/test/FailureTest.java:141]: We can return `Ray.getRuntimeContext().wasCurrentActorRestarted()` in the "ok" method and assert the result.

File: python/ray/tests/test_failure.py [hunk 5]
```diff
@@ -291,6 +292,44 @@ def fail_method(self):
...
+    errors = get_error_message(p, 1, ray_constants.TASK_PUSH_ERROR)
+    print(errors)
+    assert len(errors) == 1
+    assert errors[0].type == ray_constants.TASK_PUSH_ERROR
+    assert error_message1 in errors[0].error_message
+
+    # finally after restarting, actor will be alive
+    def wait_until_ok():
+        try:
+            return ray.get(a.ok.remote()) == "OK"
```
Comment (kfstorm) [inline python/ray/tests/test_failure.py:325]: We can return `ray.get_runtime_context().was_current_actor_reconstructed` in the "ok" method and assert the result.

**Comment by rkooo567** [comment]
To be clear if I understood, is it what's happening?

Actor created => initi failed because of random exceptions => This actor never restarts and raise init error for all actor methods?

**Comment by lixin-wei** [comment]
> To be clear if I understood, is it what's happening?
> 
> Actor created => initi failed because of random exceptions => This actor never restarts and raise init error for all actor methods?

yes, it is.

File: java/test/src/main/java/io/ray/test/FailureTest.java (hunk 4 above)
Comment (lixin-wei) [reply java/test/src/main/java/io/ray/test/FailureTest.java:141]: The "ok" method is for checking whether the actor is alive for method calls. Return a constant string is enough. I think It's not necessary to assert wasCurrentActorRestarted() in this method.

File: python/ray/tests/test_failure.py (hunk 5 above)
Comment (lixin-wei) [reply python/ray/tests/test_failure.py:322]: same as commented in java tests.

File: python/ray/tests/test_failure.py (hunk 3 above)
Comment (lixin-wei) [reply python/ray/tests/test_failure.py]: Done. Renaming was_current_actor_reconstructed is unrelated to this PR, I'll do it in another PR.

**Comment by lixin-wei** [comment]
@kfstorm All reviews are resolvedd. Please review again.

**Review by stephanie-wang — COMMENTED** [review]
I think this PR requires an API review given the changes in behavior. I'm worried about the scenario where a user will no longer receive an exception for actors that failed to be created. That seems more likely to lead to hidden bugs, since the user won't get immediate feedback if there is an issue in the actor constructor.

In general, we only retry tasks on system-level errors, not application-level errors, so this is a pretty significant break from that convention.

cc'ing @pcmoritz for comment.

File: python/ray/tests/test_actor.py [hunk 6]
```diff
@@ -296,24 +296,6 @@ def f():
     print(ray.get(x_id))  # This should not hang.
 
 
-def test_actor_init_error_propagated(ray_start_regular_shared):
```
Comment (ericl) [inline python/ray/tests/test_actor.py:299]: Hey @lixin-wei we can't break API behavior, this PR must be able to pass existing tests without changes.

**Review by ericl — CHANGES_REQUESTED** [review]

**Comment by ericl** [comment]
@lixin-wei I think there are two issues here, whether exceptions should count as errors, and preserving existing semantics.

For *whether exceptions should count as errors*: I'm not seeing a good solution here beyond adding a new flag (e.g., `treat_exceptions_as_error`).

For, preserving the original behavior, it seems simple enough to re-raise the exception after all retries are exhausted, e.g.,
1. If max_restarts=0, the existing behavior is preserved.
2. If max_restarts=N, the creation failure is only raised after N retries.

Would this satisfy the use cases outlined?

**Comment by raulchen** [comment]
@lixin-wei let's create an issue to get more comments before proceeding with this PR.

**Comment by lixin-wei** [comment]
> For whether exceptions should count as errors: I'm not seeing a good solution here beyond adding a new flag (e.g., treat_exceptions_as_error).

> In general, we only retry tasks on system-level errors, not application-level errors, so this is a pretty significant break from that convention.

@ericl @stephanie-wang For whether exceptions should count as errors, I think all exceptions in the constructor should be count as errors by default. All exceptions should make actor to restart. The convention that app-level errors in constructor won't be retried would make new users confused, we should break it. What do you think?

> I'm worried about the scenario where a user will no longer receive an exception for actors that failed to be created. That seems more likely to lead to hidden bugs, since the user won't get immediate feedback if there is an issue in the actor constructor.

> For, preserving the original behavior, it seems simple enough to re-raise the exception after all retries are exhausted

@ericl @stephanie-wang For preserving this original behavior. I agree with @ericl's opinion. By the way, in current implemention we just save the error in memory, once the actor process exits, user can't get this error anymore, instead, get an 'Actor died before executing....' exception. I think there should be a more friend way to offer this error to user. But this is not related to this PR, we can discuss it later.

So my finally solution is:
1. All exceptions in constructor should make actor to restart.
2. Re-raise the exception after all retries are exhausted as @ericl suggested.

Do you agree with this?

**Comment by lixin-wei** [comment]
> @lixin-wei let's create an issue to get more comments before proceeding with this PR.

Since we are already discussed a lot here. I think we should keep disscusing in this PR to get full context.

**Comment by ericl** [comment]
> So my finally solution is:

IIUC, this won't change behavior for actors unless max_restarts is set (since the default is zero), so the change would be a no-op for most users.

If the solution indeed only affects max_restarts > 0, then it seems ok to me. cc @stephanie-wang any concerns about this special casing for errors in the actor constructor?

**Comment by pcmoritz** [comment]
I like Eric's suggestion (making it possible to treat exceptions as errors with a flag and preserve the original behavior if max_calls = 0). The two downsides I see are the need for a new flag and also that there is no good way to get more flexibility for the application (i.e. only retrying for certain exceptions).

You can get similar behavior today without any changes to Ray by defining a metaclass that converts exceptions into system level errors:

```
def exception_as_error(func):
    import sys
    def wrapper(*args, **kwargs):
        try:
            func(*args, **kwargs)
        except:
            sys.exit(-1)
    return wrapper


class RestartOnException(type):
    def __init__(cls, classname, bases, class_dict):
        for attr_name in dir(cls):
            if attr_name == "__class__":
                # the metaclass is a callable attribute too, 
                # but we want to leave this one alone
                continue
            attr = getattr(cls, attr_name)
            if hasattr(attr, '__call__'):
                attr = exception_as_error(attr)
                setattr(cls, attr_name, attr)

        # not need for the `return` here
        super(RestartOnException, cls).__init__(classname, bases, class_dict)

@ray.remote(max_restarts=5, max_task_retries=5)
class Actor:
    __metaclass__ = RestartOnException
    def __init__(self):
        import random
        print("call __init__")
        if random.randint(a=0, b=1) == 0:
            raise Exception("Hello")
    def f(self):
        print("call")
```

Unfortunately it doesn't allow propagating the exception to the user out of the box. We could also put a utility like this into `ray.util` and use Ray's internal key value store to propagate to the user. The `__metaclass__` syntax is a little clumsy, but it has the advantage that it would be more customizable by applications (for example only restarting when certain exceptions happen).

**Comment by robertnishihara** [comment]
I agree with @stephanie-wang that this feels inconsistent with the rest of Ray, where we don't retry task when there are application exceptions.

E.g., this doesn't retry anything

```python
@ray.remote 
def f(): 
    raise Exception('error')

ray.get(f.remote())
```

but this does
```python
@ray.remote 
def f(): 
    import sys
    sys.exit(0)

ray.get(f.remote())
```

**Comment by pcmoritz** [comment]
I think if we go with Eric's suggestion we would have the same flag `treat_exceptions_as_error` for actor decorators and task decorators, which would make it consistent.

Using the `@exception_as_error` as a decorator in the solution I posted would also work for tasks.

**Comment by raulchen** [comment]
I just summarized a table for different cases: https://docs.google.com/document/d/1iDNR_oDq_WX7nq7wPAv17quLfDnqWvLTVks4Sdpv-KI/edit

Overall, I think it makes sense to treat constructor exceptions as errors. Because the actor is no longer usable, there is no point to keep it alive.
For exceptions in actor tasks, we may need more discussions. Because both behaviors make sense to some degree.

**Comment by raulchen** [comment]
My suggestion is to proceed with this PR and only kill the actor on constructor exceptions. This doesn't change the behavior when num_restarts=0. And can fix the issue that actor process still wasting resources. When num_restarts>0, I think the current behavior is invalid. So this PR fixes the issue, although there is a behavior change.

**Comment by stephanie-wang** [comment]
I would prefer to merge this change with a flag for treating exceptions as errors. The reason is that if max_restarts is set to -1 (which it often is when the flag is used), it will be impossible to receive the exception if there is a bug in the application code.

**Comment by lixin-wei** [comment]
@stephanie-wang @pcmoritz @robertnishihara  I think it's not reasonable to add flags to keep the original behavior. Since the original behavior will make users confused. Or we can add this flag but make it enabled as default.

> I would prefer to merge this change with a flag for treating exceptions as errors. The reason is that if max_restarts is set to -1 (which it often is when the flag is used), it will be impossible to receive the exception if there is a bug in the application code.

@stephanie-wang It's not our responsibility to deal with app-level bugs. The only thing we can do is to offer exception messages friendly. But this is not related to whether we should retry creation tasks.

To offer exception messages, we can use other methods such as:
1. publish events
2. add a callback when creating actors: `actor.remote(exception_callback)`

But for whether we should retry creation tasks, I stand sticky yes, because this is more reasonable and won't make user confused.

> I just summarized a table for different cases: https://docs.google.com/document/d/1iDNR_oDq_WX7nq7wPAv17quLfDnqWvLTVks4Sdpv-KI/edit

I think @raulchen's table is very nice. The excepted behavior, when exceptions raised in the constructor, is to kill this actor and restart. That's exactly what new users would think.

**Comment by lixin-wei** [comment]
> I agree with @stephanie-wang that this feels inconsistent with the rest of Ray, where we don't retry task when there are application exceptions.

> I think if we go with Eric's suggestion we would have the same flag treat_exceptions_as_error for actor decorators and task decorators, which would make it consistent.

As for inconsistency between actor creation task and normal tasks/actor method tasks. I think it's OK to keep this inconsistency because it's intuitive to the user. Normal task and actor method task should not be retried, while actor creation task should be retried.

**Comment by lixin-wei** [comment]
> I agree with @stephanie-wang that this feels inconsistent with the rest of Ray, where we don't retry task when there are application exceptions.

> I think if we go with Eric's suggestion we would have the same flag treat_exceptions_as_error for actor decorators and task decorators, which would make it consistent.

As for inconsistency between actor creation task and normal tasks/actor method tasks. I think it's OK to keep this inconsistency because it's intuitive to the user. Normal task and actor method task should not be retried, while actor creation task should be retried when max_restarts is set. As @raulchen's table shows.

**Comment by lixin-wei** [comment]
In conclusion, I think we don't need to add `treate_exception_as_errors` flag. Instead, make it as default in creation tasks.

**Comment by raulchen** [comment]
> I would prefer to merge this change with a flag for treating exceptions as errors. The reason is that if max_restarts is set to -1 (which it often is when the flag is used), it will be impossible to receive the exception if there is a bug in the application code.

@stephanie-wang What about just publishing an error to error table every time when the actor restarts? We can include the app-level exception in the error message. 

I'd prefer not adding a flag for constructors.

**Comment by ericl** [comment]
Could we raise the exception to previous callers of the actor only? That
will provide feedback even in the case of -1 restarts.

**Comment by lixin-wei** [comment]
> Could we raise the exception to previous callers of the actor only? That will provide feedback even in the case of -1 restarts.
@ericl The 'caller' of creation task is not a sync call. Do you mean adding a callback to `Actor.remote()` like `Actor.remote(exception_callback)`, then pass exceptions by this callback?

**Comment by ericl** [comment]
The exceptions can be raised against further method calls right, like in
the current implementation right?

So we can fail the method calls until the actor is successfully restarted.
If the actor is failing to restart forever, all method calls would be
raising exceptions forever.

Example:
a = Actor.remote() # first construction fails
ray.get(a.f.remote()) # raises error
sleep(1)
ray.get(a.f.remote()) # actor restarted, no error

**Comment by lixin-wei** [comment]
> The exceptions can be raised against further method calls right, like in the current implementation right? So we can fail the method calls until the actor is successfully restarted. If the actor is failing to restart forever, all method calls would be raising exceptions forever. Example: a = Actor.remote() # first construction fails ray.get(a.f.remote()) # raises error sleep(1) ray.get(a.f.remote()) # actor restarted, no error

I think this makes sense. For implementation, we can save exception in GCS and publish it to all workers.

To sum up:
1. restart actor once exception raised in the constructor, no flags.
2. raise constructor exceptions to callers until the actor is alive.

Do you all agree with this? @ericl @stephanie-wang @raulchen @robertnishihara @pcmoritz

**Comment by lixin-wei** [comment]
But hold on, please. After discussing with @raulchen and @kfstorm, we think it's better to mark actor as DEAD immediately when an exception raised in constructor.  Because in normal programming, when you have exceptions in class's constructor, this class will not be created successfully. We want to keep this consistency.

We are talking with our users to check if they can accept this behavior. If they do, I'll close this PR and open another PR to make actor DEAD immediately after exceptions in the constructor.

**Comment by stephanie-wang** [comment]
Since this PR thread is getting pretty long, could we move the discussion to a GitHub issue and request comments from the community through the Ray slack?

**Comment by lixin-wei** [comment]
> Since this PR thread is getting pretty long, could we move the discussion to a GitHub issue and request comments from the community through the Ray slack?

OK, and our users accept the solution I mentioned above. I'll close this PR and move to another.

**Pull request closed without merging**